src/core/ddsc/src/dds_topic.c

```diff
@@ -78,24 +78,18 @@ static dds_return_t dds_topic_status_validate (uint32_t mask)
 #ifdef DDS_HAS_TOPIC_DISCOVERY
 static struct ktopic_type_guid * topic_guid_map_refc_impl (const struct dds_ktopic * ktp, const struct ddsi_sertype *sertype, bool unref)
 {
-  type_identifier_t *tid = ddsi_typeid_from_sertype (sertype);
-  if (ddsi_typeid_none (tid))
-  {
-    // tid may be a null pointer but it may also be all-zero
-    ddsrt_free (tid);
+  ddsi_typeid_t *type_id = ddsi_sertype_typeid (sertype, DDSI_TYPEID_KIND_COMPLETE);
+  if (ddsi_typeid_is_none (type_id))
     return NULL;
```

Comment:
The type id is returned by a sertype implementation, so it's dangerous to assume that the `DDS_XTypes_TK_NONE` case cannot occur. I'll replaced the check by `ddsi_typeid_is_none (type_id) || !ddsi_typeid_is_hash (type_id)` (because a sertype's typeid must be a hashed id) and free the type id also in case it is null or non-hash.